packages/IntegrationTest/windows/integrationtest/TestHostHarness.cpp

```diff
@@ -0,0 +1,195 @@
+using namespace std::chrono_literals;
+
+using namespace winrt::integrationtest;
+using namespace winrt::Microsoft::ReactNative;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Data::Json;
+
+namespace IntegrationTest {
+
+TestHostHarness::TestHostHarness() noexcept {
+  m_redboxHandler = winrt::make<TestHostHarnessRedboxHandler>(get_weak());
+  m_commandListener = winrt::make_self<TestCommandListener>();
+
+  m_commandListener->OnTestCommand([weakThis{get_weak()}](
+      const TestCommand &command, TestCommandResponse response) noexcept {
+    if (auto strongThis = weakThis.get()) {
+      strongThis->OnTestCommand(command, std::move(response));
+    }
+  });
+}
+
+void TestHostHarness::SetRootView(ReactRootView &&rootView) noexcept {
+  m_rootView = std::move(rootView);
+}
+
+void TestHostHarness::SetReactHost(ReactNativeHost &&reactHost) noexcept {
+  reactHost.InstanceSettings().RedBoxHandler(m_redboxHandler);
+  m_instanceLoadedRevoker = reactHost.InstanceSettings().InstanceLoaded(
+      winrt::auto_revoke, [weakThis{get_weak()}](const auto & /*sender*/, const auto &args) noexcept {
+        auto strongThis = weakThis.get();
+        if (!strongThis) {
+          return;
+        }
+
+        ReactContext context(args.Context());
+
+        strongThis->m_context = context;
+        strongThis->StartListening();
+
+        context.Notifications().Subscribe(
+            TestModule::TestCompletedEvent(),
+            context.UIDispatcher(),
+            [weakThis](const auto & /*sender*/, ReactNotificationArgs<void> /*args*/) noexcept {
+              if (auto strongThis = weakThis.get()) {
+                strongThis->OnTestCompleted();
+              }
+            });
+
+        context.Notifications().Subscribe(
+            TestModule::TestPassedEvent(),
+            context.UIDispatcher(),
+            [weakThis](const auto & /*sender*/, ReactNotificationArgs<bool> args) noexcept {
+              if (auto strongThis = weakThis.get()) {
+                strongThis->OnTestPassed(*args.Data());
+              }
+            });
+      });
+}
+
+winrt::fire_and_forget TestHostHarness::StartListening() noexcept {
+  auto listenResult = co_await m_commandListener->StartListening();
+
+  switch (listenResult) {
+    case ListenResult::Success:
+      co_return;
+
+    case ListenResult::AddressInUse:
+      ShowJSError("Test harness address is already in use. Is there another instance running?");
+      co_return;
+
+    case ListenResult::OtherError:
+      ShowJSError("The test harness ran into an unexpected error trying to listen for commands.");
+      co_return;
+
+    default:
+      ShowJSError("Unexpected ListenResult from TestCommandListener.");
+      co_return;
+  }
+}
+
+winrt::fire_and_forget TestHostHarness::OnTestCommand(TestCommand command, TestCommandResponse response) noexcept {
+  // Keep ourselves alive while we have a test command
+  auto strongThis = get_strong();
```

Comment:
Should figure out something sane to do if we get a test command when a previous is still active.